pyccel/ast/core.py

```diff
@@ -3291,8 +3291,25 @@ def name(self):
 
     @property
     def attributes(self):
+        """
+        The attributes of a class.
+
+        A tuple of the attributes of a ClassDef.
+
+        Returns
+        -------
+        None
+        """
+
         return self._attributes
 
+    @attributes.setter
+    def attributes(self, attr):
+        if not isinstance(attr, tuple):
+            raise TypeError('attributes must be a tuple.')
+        self._attributes = attr
+        self.set_current_user_node(self)
```

Comment:
```suggestion
    @attributes.setter
    def attributes(self, attr):
        if not isinstance(attr, tuple):
            raise TypeError('attributes must be a tuple.')
        for a in self._attributes:
            self.remove_user_node(a)
        self._attributes = attr
        for a in attr:
             a.set_current_user_node(self)
```

Reply:
I wonder if it would make more sense to have a simpler `add_new_attribute`:
```python
def add_new_attribute(self, attr):
     if not isinstance(attr, Variable):
          raise TypeError("Attributes must be Variables")
     attr.set_current_user_node(self)
     self._attributes += (attr,)
```

Reply:
If we want to handle name collisions properly we should probably also add new attributes to the class scope:
```python
self.scope.insert_variable(attr)
```

Reply:
Fixed.

Reply:
we don't really need the `add_new_attribute` the adding of the new attributes is already handled in : https://github.com/pyccel/pyccel/blob/58a355e5f498b271a4c456478a397c40c0d4f9fa/pyccel/parser/semantic.py#L1165-L1169

Reply:
I used it here so we can avoid the creating of the class each time
https://github.com/pyccel/pyccel/blob/58a355e5f498b271a4c456478a397c40c0d4f9fa/pyccel/parser/semantic.py#L1170-L1174

Reply:
Then I think this is a problem. For example:
```python
1 class MyClass:
2     def __init__(self, a : int, b : float, c : int):
3          self.t = (a,b)
4          self.c = c
```
At line 3 we create an `InhomogeneousTupleVariable` (`t`). This variable uses 2 sub-variables `t_0001` and `t_0002` which represent respectively the `int` and `float` parts of the tuple.
We create a class and save the attribute in this class.

At line 4 we create a `Variable`. We need to add this to the attributes of the class so we call the attribute setter.
The setter is designed to completely replace all the attributes with a new set of unrelated attributes so it begins by
telling the attributes that it currently uses that they are no longer needed.
The `InhomogeneousTupleVariable` `t` sees that it is no longer used by anyone. It assumes that this means it will never be used. It therefore no longer needs to use anything else and should be removed from the `users` of anything it is using so they do not have the wrong tree (this makes more sense in the case of something like a `PyccelAdd` which you may decide is unnecessary. You want it to be garbage collected, but it can't be as long as it says it is using existing Variables as they store it as an attribute). It therefore removes itself from the users of `t_0001` and `t_0002`.
Now we add the new attributes.
We therefore tell them that they have new users. `t` is also in the new users, so it is not garbage collected, however it is no longer aware that it has the attributes `t_0001` and `t_0002` which could cause big problems later.

There are therefore 2 solutions:
1. Add an `invalidate` argument to the setter (which will therefore become a function) so you don't invalidate the arguments if you are planning to reuse them.
2. Stop doing so many unnecessary steps and just have a `add_new_attribute` function

Reply:
Fixed.